Added non-digested version for assets (images, fonts, svg)

alleycat-at-git opened the pull request:
**Problem**
When client js uses images in render methods, e.g. `<img src='...' />` or in css, e.g. `background-image: url(...)` these assets fails to load. This happens because rails adds digest hashes to filenames when compiling assets. e.g. `img1.jpg` becomes `img1-dbu097452jf2v2.jpg`. When compiling its native css rails transforms all urls and links to work on digested versions, i.e. `background-image: url(img1.jpg)` becomes `background-image: url(img1-dbu097452jf2v2.jpg)`. However this doesn't happen for js and css files compiled by webpack on the client side and assets fail to load

**Solution**

Create symlinks of non-digested versions to digested versions when rails assets compile

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/shakacode/react_on_rails/413)
<!-- Reviewable:end -->

justin808 commented:
Let's look at this sample: https://github.com/shakacode/react-webpack-rails-tutorial/blob/master/lib/tasks/assets.rake

I like the idea of putting these in standard tasks for the gem. However, we need to consider the upgrade path for existing ReactOnRails apps. For example, somebody's app might already be doing this, so we need the ability to configure if this done.

Next, we should consider these below (from the link above). Maybe we can prefix the tasks we offer with the namespace `react_on_rails::`.

This is super awesome! Let's do this once. We'll also need to update the docs. That's actually as or more important than the generator.

CC: @jbhatab @robwise @aaronvb 



```ruby
Rake::Task["assets:precompile"]
  .clear_prerequisites
  .enhance(["assets:compile_environment"])

namespace :assets do
  # In this task, set prerequisites for the assets:precompile task
  task compile_environment: :webpack do
    Rake::Task["assets:environment"].invoke
  end

  desc "Compile assets with webpack"
  task :webpack do
    sh "cd client && npm run build:production:client"
    sh "cd client && npm run build:production:server"
    sh "mkdir -p public/assets"

    # Critical to manually copy non js/css assets to public/assets as we don't want to fingerprint them
    sh "cp -rf app/assets/webpack/*.woff* app/assets/webpack/*.svg app/assets/webpack/*.ttf "\
       "app/assets/webpack/*.eot* public/assets"

    # TODO: We should be gzipping these
  end

  task :clobber do
    rm_rf "#{Rails.application.config.root}/app/assets/webpack"
  end
end
```

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-:-KGx0luAd5qiZTsz5sao:897042032)*
<!-- Sent from Reviewable.io -->

alleycat-at-git commented:
@justin808 I refactored it to make a hook inside React-on-rails gem. Now we don't have these problems. Also added docs and changelog

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-:-KGxCXoLwVzzIxnPd-MB:-78255794)*
<!-- Sent from Reviewable.io -->

justin808 commented:
Added more comments.

We need to test this as a migration to https://github.com/shakacode/react-webpack-rails-tutorial/

---

Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[docs/additional-reading/rails-assets.md, line 19 \[r2\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KGxIHhUZ093fcYXwpOK:-KGxIHhVbuya24Wr4PST:-1926113239) ([raw file](https://github.com/shakacode/react_on_rails/blob/9f667aa17c55091497c471d57dc0ea22a9fe8113/docs/additional-reading/rails-assets.md#L19)):*
Will the default value be nil or otherwise? if this is not specified. Does this require a major version bump?

---

*[lib/react_on_rails/configuration.rb, line 60 \[r1\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KGxIRGOb7kCayWIM3oh:-KGxIRGPw8KfCfFZa1KE:1376610399) ([raw file](https://github.com/shakacode/react_on_rails/blob/5c59be4daf63d5558e91732884b457ddb253940e/lib/react_on_rails/configuration.rb#L60)):*
OK -- default to nil, so this is not done. It's opt-in.

---

*[lib/tasks/assets.rake, line 2 \[r2\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KGxIcDyA0emAT0R8rFN:-KGxIcDyA0emAT0R8rFO:153921966) ([raw file](https://github.com/shakacode/react_on_rails/blob/9f667aa17c55091497c471d57dc0ea22a9fe8113/lib/tasks/assets.rake#L2)):*
shouldn't we be checking if the param is nil, so this is not done?

---

*[lib/generators/react_on_rails/templates/base/base/lib/tasks/assets.rake.tt, line 27 \[r1\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KGxIiSZyYDqdoTVSWPd:-KGxIiSZyYDqdoTVSWPe:1517154141) ([raw file](https://github.com/shakacode/react_on_rails/blob/5c59be4daf63d5558e91732884b457ddb253940e/lib/generators/react_on_rails/templates/base/base/lib/tasks/assets.rake.tt#L27)):*
looks like we're duplicating this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-:-KGxIoGieFsqYy0NvDWr:1450931840)*
<!-- Sent from Reviewable.io -->

alleycat-at-git commented:
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[docs/additional-reading/rails-assets.md, line 19 \[r2\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KGxIHhUZ093fcYXwpOK:-KGzd4iURqvU0bfB2OPr:2138603896) ([raw file](https://github.com/shakacode/react_on_rails/blob/9f667aa17c55091497c471d57dc0ea22a9fe8113/docs/additional-reading/rails-assets.md#L19)):*
I guess I will default it to images, fonts and svgs. It doesn't hurt anything and totally backwards-compatible.

---

*[lib/react_on_rails/configuration.rb, line 60 \[r1\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KGxIRGOb7kCayWIM3oh:-KGzdQre_flVveoI8dkd:1938135954) ([raw file](https://github.com/shakacode/react_on_rails/blob/5c59be4daf63d5558e91732884b457ddb253940e/lib/react_on_rails/configuration.rb#L60)):*
Will change that to to images, fonts and svgs

---

*[lib/tasks/assets.rake, line 2 \[r2\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KGxIcDyA0emAT0R8rFN:-KGzdVY6qlDPFmDHNuOd:2071384388) ([raw file](https://github.com/shakacode/react_on_rails/blob/9f667aa17c55091497c471d57dc0ea22a9fe8113/lib/tasks/assets.rake#L2)):*
Yes! Nice comment!

---

*[lib/generators/react_on_rails/templates/base/base/lib/tasks/assets.rake.tt, line 27 \[r1\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KGxIiSZyYDqdoTVSWPd:-KH-3bL6mTTV2Hk9OXxJ:1023586464) ([raw file](https://github.com/shakacode/react_on_rails/blob/5c59be4daf63d5558e91732884b457ddb253940e/lib/generators/react_on_rails/templates/base/base/lib/tasks/assets.rake.tt#L27)):*
Can you explain what do we duplicate here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/413)*
<!-- Sent from Reviewable.io -->

justin808 commented:
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[lib/generators/react_on_rails/templates/base/base/lib/tasks/assets.rake.tt, line 27 \[r1\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KGxIiSZyYDqdoTVSWPd:-KH-611Bp2g_FIqD5mSo:170409872) ([raw file](https://github.com/shakacode/react_on_rails/blob/5c59be4daf63d5558e91732884b457ddb253940e/lib/generators/react_on_rails/templates/base/base/lib/tasks/assets.rake.tt#L27)):*
this is the generator -- why are we creating this task in the generated app when it seems to be in the gem

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/413)*
<!-- Sent from Reviewable.io -->

alleycat-at-git commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[lib/generators/react_on_rails/templates/base/base/lib/tasks/assets.rake.tt, line 27 \[r1\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KGxIiSZyYDqdoTVSWPd:-KH-hi2m-1LIV9zYMWqK:1833726019) ([raw file](https://github.com/shakacode/react_on_rails/blob/5c59be4daf63d5558e91732884b457ddb253940e/lib/generators/react_on_rails/templates/base/base/lib/tasks/assets.rake.tt#L27)):*
Symlinking task is almost 100% configurable via ReactOnRails.config, this task however cannot be configured that easy, so I'd rather leave it here - this way framework user could easily override it. But if you think it makes sense to move it to the gem, I can do that. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/413)*
<!-- Sent from Reviewable.io -->

robwise commented:
Reviewed 4 of 5 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[lib/react_on_rails/configuration.rb, line 60 \[r3\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KH169Kf4jmNG4WEblv1:-KH13jnpu5x_xH7R_RPb:-1841572230) ([raw file](https://github.com/shakacode/react_on_rails/blob/af105674ee9071b3a4e5e7515201c1d4be3c52fc/lib/react_on_rails/configuration.rb#L60)):*
Rails already creates non-digest versions of `svg`s, `eot`s, `woff`s, and `ttf`s, doesn't it? It's just the picture types?

---

*[lib/tasks/assets.rake, line 11 \[r3\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KH151S9n5lM2lTh6YhT:-KH13cnX7S1zb0jkYTYk:-493215331) ([raw file](https://github.com/shakacode/react_on_rails/blob/af105674ee9071b3a4e5e7515201c1d4be3c52fc/lib/tasks/assets.rake#L11)):*
@alleycat-at-git will this also pick up the gzipped assets and provide non-fingerprinted versions of that? We need to make sure we make gzipped versions of the non-digested assets as well for the server to use.

---

*[lib/tasks/assets.rake, line 15 \[r3\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KH141VzyyTuRoZ_C8ID:-KH141VzyyTuRoZ_C8IE:-42861450) ([raw file](https://github.com/shakacode/react_on_rails/blob/af105674ee9071b3a4e5e7515201c1d4be3c52fc/lib/tasks/assets.rake#L15)):*
minor thing, but it's less typing to do `Rails.root.join('public/assets', logical_path)`

---

*[lib/tasks/assets.rake, line 17 \[r3\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KH14ikPLMW2Acd5Q1J4:-KH14ikQqZ3l6bOLS4-q:-43788502) ([raw file](https://github.com/shakacode/react_on_rails/blob/af105674ee9071b3a4e5e7515201c1d4be3c52fc/lib/tasks/assets.rake#L17)):*
we're sure these will get wiped every time we recompile, right? 

Like what if:

- we put an image in our client build
- webpack outputs the image to `app/assets/webpack`
- rails compiles and places digested version of this image in `public/assets`
- we create a non-digested symlink to this file in `public/assets`
- we delete the source image in our client build, so webpack no longer outputs it to `app/assets/webpack`
- rails compiles and does not place any image in the `public/assets` folder

would the non-digested symlink we made now be invalid and pointing to nowhere? Or will it have gotten deleted?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/413)*
<!-- Sent from Reviewable.io -->

alleycat-at-git commented:
Review status: 4 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[lib/react_on_rails/configuration.rb, line 60 \[r3\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KH169Kf4jmNG4WEblv1:-KH1koBLFP4AkeVdLRtZ:-995214027) ([raw file](https://github.com/shakacode/react_on_rails/blob/af105674ee9071b3a4e5e7515201c1d4be3c52fc/lib/react_on_rails/configuration.rb#L60)):*
I guess it digests everything - here's the assets of a brand new rails project
123-e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855.svg
123-e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855.svg.gz
123-e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855.woff
application-e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855.css
application-e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855.css.gz
application-f51a52e2dc61278f78b2b181286b91096cf236e0942e6f57f5c5b8dc24fa28f7.js
application-f51a52e2dc61278f78b2b181286b91096cf236e0942e6f57f5c5b8dc24fa28f7.js.gz


Anyway the comment was useful, thanks! I updated the method to be more cautious and overwrite only symlinks

---

*[lib/tasks/assets.rake, line 11 \[r3\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KH151S9n5lM2lTh6YhT:-KH1qsQcgdVjzYYlks6R:717155465) ([raw file](https://github.com/shakacode/react_on_rails/blob/af105674ee9071b3a4e5e7515201c1d4be3c52fc/lib/tasks/assets.rake#L11)):*
Oh yes, definitely - the way it is used we also symlink all gz versions as we don't specify the end of line in the regex

---

*[lib/tasks/assets.rake, line 15 \[r3\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KH141VzyyTuRoZ_C8ID:-KH1n8LoKkgMvHFPIMLs:-499146009) ([raw file](https://github.com/shakacode/react_on_rails/blob/af105674ee9071b3a4e5e7515201c1d4be3c52fc/lib/tasks/assets.rake#L15)):*
Done.

---

*[lib/tasks/assets.rake, line 17 \[r3\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KH14ikPLMW2Acd5Q1J4:-KH1pF2HePgbd6Mgdf5u:744313620) ([raw file](https://github.com/shakacode/react_on_rails/blob/af105674ee9071b3a4e5e7515201c1d4be3c52fc/lib/tasks/assets.rake#L17)):*
In this case the image despite being removed from webpack will still reside in public/assets. What we can do though is check for broken symlinks each time we recompile the assets.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/413)*
<!-- Sent from Reviewable.io -->

alleycat-at-git commented:
Review status: 4 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[lib/tasks/assets.rake, line 11 \[r3\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KH151S9n5lM2lTh6YhT:-KH2GuQrrqz3smYUvy37:1090963993) ([raw file](https://github.com/shakacode/react_on_rails/blob/af105674ee9071b3a4e5e7515201c1d4be3c52fc/lib/tasks/assets.rake#L11)):*
Wrong comment - actually I added the code for gz versions as well

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/413)*
<!-- Sent from Reviewable.io -->

alleycat-at-git commented:
@justin808 I checked this PR with react-webpack-rails-tutorial and it totally works. Though we should correct it a bit to remove clutter after we merge this PR

---

Review status: 4 of 5 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-:-KH2N3AWVuL-tWxP3VF-:1707794524)*
<!-- Sent from Reviewable.io -->

robwise commented:
@alleycat-at-git make sure you check with an image that is large enough that webpack won't just inline it, that's how we never caught this before

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  good job

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[lib/react_on_rails/configuration.rb, line 60 \[r3\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KH169Kf4jmNG4WEblv1:-KH2TbxUsfmux-tGwI_e:592896086) ([raw file](https://github.com/shakacode/react_on_rails/blob/af105674ee9071b3a4e5e7515201c1d4be3c52fc/lib/react_on_rails/configuration.rb#L60)):*
oh okay

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-:-KH2TXP1NHfBgDL4Ug0r:-1557225664)*
<!-- Sent from Reviewable.io -->

alleycat-at-git commented:
@robwise Thanks! )

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-:-KH3eXiRcpPz0TZaCgtk:2008621679)*
<!-- Sent from Reviewable.io -->

justin808 commented:
I had a suggestion and question.

Looks *AMAZING*!

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/additional-reading/rails-assets.md, line 19 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KH3sAIl6gJtyR691emG:-KH3sAIl6gJtyR691emH:902464858) ([raw file](https://github.com/shakacode/react_on_rails/blob/fc23f2e27a5b1d70f1cf82973f0972bd7fbc0f7d/docs/additional-reading/rails-assets.md#L19)):*
Are symlinks definitely better than just copying the assets on deploy to the given names?

---

*[lib/tasks/assets.rake, line 22 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KH3rbZc_qI_Pco4oJR2:-KH3rbZc_qI_Pco4oJR3:939637661) ([raw file](https://github.com/shakacode/react_on_rails/blob/fc23f2e27a5b1d70f1cf82973f0972bd7fbc0f7d/lib/tasks/assets.rake#L22)):*
```ruby
if ReactOnRails.configuration.symlink_non_digested_assets_regex.present?
```

otherwise, if this is set to a "", then that's truthy

---

*[lib/generators/react_on_rails/templates/base/base/lib/tasks/assets.rake.tt, line 27 \[r1\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KGxIiSZyYDqdoTVSWPd:-KH3rtXkN-7pnnlPOvwl:1402398076) ([raw file](https://github.com/shakacode/react_on_rails/blob/5c59be4daf63d5558e91732884b457ddb253940e/lib/generators/react_on_rails/templates/base/base/lib/tasks/assets.rake.tt#L27)):*
Yes, let's depend on the gem. If somebody needs to change it, they can find the source and copy it.

Otherwise, this is just extra stuff that @jbhatab will have to explain when he creates a video of how the generators work.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-:-KH3sFuPWHWOMD8aBeli:987636745)*
<!-- Sent from Reviewable.io -->

alleycat-at-git commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/additional-reading/rails-assets.md, line 19 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KH3sAIl6gJtyR691emG:-KH3uLGXH2vMsjJaSMWQ:-1008130139) ([raw file](https://github.com/shakacode/react_on_rails/blob/fc23f2e27a5b1d70f1cf82973f0972bd7fbc0f7d/docs/additional-reading/rails-assets.md#L19)):*
I think yes - this just saves some space and ensures that we have absolutely identical assets. 

---

*[lib/tasks/assets.rake, line 22 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KH3rbZc_qI_Pco4oJR2:-KH3v1VzLYXfSYF9pspM:1943076078) ([raw file](https://github.com/shakacode/react_on_rails/blob/fc23f2e27a5b1d70f1cf82973f0972bd7fbc0f7d/lib/tasks/assets.rake#L22)):*
This is not a string, but a regex, so if you make it // then it will actually match any asset

---

*[lib/generators/react_on_rails/templates/base/base/lib/tasks/assets.rake.tt, line 27 \[r1\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KGxIiSZyYDqdoTVSWPd:-KH3vZligc9BVwIiEF1X:5173697) ([raw file](https://github.com/shakacode/react_on_rails/blob/5c59be4daf63d5558e91732884b457ddb253940e/lib/generators/react_on_rails/templates/base/base/lib/tasks/assets.rake.tt#L27)):*
Ok

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/413)*
<!-- Sent from Reviewable.io -->

alleycat-at-git commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[lib/generators/react_on_rails/templates/base/base/lib/tasks/assets.rake.tt, line 27 \[r1\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/413#-KGxIiSZyYDqdoTVSWPd:-KH3wF7te24olfSPavgG:-95191778) ([raw file](https://github.com/shakacode/react_on_rails/blob/5c59be4daf63d5558e91732884b457ddb253940e/lib/generators/react_on_rails/templates/base/base/lib/tasks/assets.rake.tt#L27)):*
@justin808 JIC this is not my code - it was there before this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/413)*
<!-- Sent from Reviewable.io -->

justin808 commented:
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/413)*
<!-- Sent from Reviewable.io -->
